File: 2_Two_Pointers/containerWithMostWater.js
var maxArea = function (height) {
    // Initialize left, right pointers, and max area.
    let [left, right, max] = [0, height.length - 1, 0];

    // Continue the loop until the left pointer is less than the right pointer.
    while (left < right) {
        // Variables to store container height and area.
        let containerHeight, area;

        // Calculate the width of the container between left and right pointers.
        let containerWidth = right - left;

        // Determine the smaller height of the two walls.
        if (height[left] < height[right]) {
            containerHeight = height[left];
            // Move the left pointer to the right.
            left++;
        } else {
            containerHeight = height[right];
            // Move the right pointer to the left.
            right--;
        }

        // Calculate the area of the current container.
        area = containerWidth * containerHeight;

        // Update the maximum area if the current area is greater.
        if (area > max) {
            max = area;
        }
    }

    // Return the maximum area found.
    return max;
};

// implementation

console.log("Leetcode 15: Container With Most Water\n");

// Eaxmple 1
example1 = [1, 8, 6, 2, 5, 4, 8, 3, 7];
console.log(`Example 1: height = [${example1}]`);
console.log("Output:", maxArea(example1));

// Example 2
example2 = [1, 1];
console.log(`Example 2: height = [${example2}]`);
console.log("Output:", maxArea(example2));

File: 2_Two_Pointers/container_with_most_water.py
class Solution:
    def maxArea(self, height: list[int]) -> int:
        # Initialize left and right pointers at the two ends of the height list.
        left, right = 0, len(height) - 1
        # Initialize the variable to store the maximum area.
        result = 0

        # Continue the loop until the left pointer is less than the right pointer.
        while left < right:
            # Calculate the current area between the two pointers and update the result.
            result = max(result, min(height[left], height[right]) * (right - left))
            
            # Move the pointers towards each other based on the height of the walls.
            if height[left] < height[right]:
                left += 1
            elif height[right] <= height[left]:
                right -= 1
            
        # Return the maximum area found.
        return result

# implementation 
answer = Solution()

print("Leetcode 15: Container With Most Water\n")

# Example 1
example_1 = [1, 8, 6, 2, 5, 4, 8, 3, 7]
print(f"Example 1: height = {example_1}")
print(f"Output:", answer.maxArea(example_1))

# Example 2
example_2 = [1, 1]
print(f"Example 2: height = {example_2}")
print("Output:", answer.maxArea(example_2))